lib/ts/controllers/s-uploader.ts

```diff
@@ -0,0 +1,171 @@
+        data?: string | ArrayBuffer;
+        name: string;
+        type: string;
+    };
+
+    export class UploaderController extends Stacks.StacksController {
+        static targets = ["input", "previews", "uploader"];
+        private inputTarget!: HTMLInputElement;
+        private previewsTarget!: HTMLElement;
+        private uploaderTarget!: HTMLElement;
+
+        private boundDragEnter!: any;
+        private boundDragLeave!: any;
+
+        private static readonly FILE_DISPLAY_LIMIT = 10;
+        private static readonly MAX_FILE_SIZE = 1024 * 1024 * 10; // 10 MB
+
+        connect() {
+            super.connect();
+            this.boundDragEnter = this.handleUploaderActive.bind(this, true);
+            this.boundDragLeave = this.handleUploaderActive.bind(this, false);
+
+            this.inputTarget.addEventListener("dragenter", this.boundDragEnter);
+            this.inputTarget.addEventListener("dragleave", this.boundDragLeave);
+        }
+
+        disconnect() {
+            this.inputTarget.removeEventListener("dragenter", this.boundDragEnter);
+            this.inputTarget.removeEventListener("dragleave", this.boundDragLeave);
+            super.disconnect();
+        }
+
+        /**
+         * Handles rendering the file preview state on input change
+         */
+        handleInput() {
+            this.previewsTarget.innerHTML = "";
+
+            if (!this.inputTarget.files) {
+                return;
+            }
+
+            const count = this.inputTarget.files.length;
+                this.getDataURLs(this.inputTarget.files, UploaderController.FILE_DISPLAY_LIMIT)
+                    .then((res) => {
+                        this.handleVisible(true);
+                        const hasMultipleFiles = res.length > 1;
+
+                        if (hasMultipleFiles) {
+                            let headingElement = document.createElement("div");
+                            headingElement.classList.add("s-uploader--previews-heading");
+                            headingElement.innerText = res.length < count ?
+                                `Showing ${res.length} of ${count} files` : `${count} items`;
+                            this.previewsTarget.appendChild(headingElement);
+                            this.previewsTarget.classList.add("has-multiple");
+                        } else {
+                            this.previewsTarget.classList.remove("has-multiple");
+                        }
+                        res.forEach((file) => this.addFilePreview(file));
+                        this.handleUploaderActive(true);
+                    });
+        }
+
+        /**
+         * Resets the Uploader to initial state
+         */
+        reset() {
+            this.inputTarget.value = '';
+            this.previewsTarget.innerHTML = "";
+            this.handleVisible(false);
+        }
+
+        /**
+         * @param  {boolean} shouldPreview - Uploader is entering a preview state
```

Comment:
nit: To me, the most important part of a function's docs is the description, followed by the param descriptions (followed by types if not using TS). I recommend adding docs for all functions (especially the public ones)